plugins/wacom-usb/fu-wac-module-scaler.c

```diff
@@ -0,0 +1,190 @@
+} FuWacModuleScalerBlockData;
+
+static guint8
+fu_wac_module_scaler_calculate_crc(const guint8 *data, gsize sz)
+{
+	guint8 crc = ~fu_crc8(data, sz);
+	return crc;
+}
+
+static GPtrArray *
+fu_wac_module_scaler_parse_blocks(const guint8 *data, gsize sz, GError **error)
+{
+	GPtrArray *blocks = g_ptr_array_new_with_free_func(g_free);
+	int i = 0;
+	for (guint addr = 0x0; addr < sz; addr += FU_WAC_MODULE_SCALER_PAYLOAD_SZ) {
+		FuWacModuleScalerBlockData *bd;
+		gsize cdata_sz = FU_WAC_MODULE_SCALER_PAYLOAD_SZ;
+
+		bd = g_new0(FuWacModuleScalerBlockData, 1);
+
+		bd->addr[0] = (addr >> 16) & 0xff;
+		bd->addr[1] = (addr >> 8) & 0xff;
+		bd->addr[2] = addr & 0xff;
+
+		memset(bd->cdata, 0xff, FU_WAC_MODULE_SCALER_PAYLOAD_SZ);
+
+		if (addr + FU_WAC_MODULE_SCALER_PAYLOAD_SZ >= sz)
+			cdata_sz = sz - addr;
+
+		if (!fu_memcpy_safe(bd->cdata,
+				    sizeof(bd->cdata),
+				    0x0, /* dst */
+				    data,
+				    sz,
+				    addr, /* src */
+				    cdata_sz,
+				    error))
+			return NULL;
+		bd->crc =
+		    fu_wac_module_scaler_calculate_crc(bd->cdata, FU_WAC_MODULE_SCALER_PAYLOAD_SZ);
+		g_ptr_array_add(blocks, bd);
+		i = i + 1;
+	}
+	return blocks;
+}
+
+static gboolean
+fu_wac_module_scaler_write_firmware(FuDevice *device,
+				    FuFirmware *firmware,
+				    FuProgress *progress,
+				    FwupdInstallFlags flags,
+				    GError **error)
+{
+	FuWacModule *self = FU_WAC_MODULE(device);
+	const guint8 *data;
+	gsize len = 0;
+	const guint8 buf_start[] = {FU_WAC_MODULE_SCALER_START_NORMAL};
+
+	g_autoptr(GPtrArray) blocks = NULL;
+	g_autoptr(GBytes) blob_start = g_bytes_new_static(buf_start, 1);
+	g_autoptr(GBytes) fw = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 8, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 59, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 33, NULL);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* build each data packet */
+	data = g_bytes_get_data(fw, &len);
+
+	blocks = fu_wac_module_scaler_parse_blocks(data, len, error);
+	if (blocks == NULL)
+		return FALSE;
+
+	/* start, which will erase the module */
+	if (!fu_wac_module_set_feature(self,
+				       FU_WAC_MODULE_COMMAND_START,
+				       blob_start,
+				       fu_progress_get_child(progress),
+				       FU_WAC_MODULE_ERASE_TIMEOUT,
+				       error))
+		return FALSE;
+
+	fu_progress_step_done(progress);
+
+	/* data */
+	for (guint i = 0; i < blocks->len; i++) {
+		FuWacModuleScalerBlockData *bd = g_ptr_array_index(blocks, i);
+		guint8 buf[FU_WAC_MODULE_SCALER_PAYLOAD_SZ + 4];
+		g_autoptr(GBytes) blob_chunk = NULL;
+
+		/* build data packet */
+		memset(buf, 0xff, sizeof(buf));
+		memcpy(&buf[0], bd->addr, 3);
+		buf[3] = bd->crc;
+		memcpy(&buf[4], bd->cdata, sizeof(bd->cdata));
+		blob_chunk = g_bytes_new(buf, sizeof(buf));
```

Comment:
I'm a bit confused why you can't do `g_bytes_new(bd, sizeof(*bd))` -- I guess it would mean marking FuWacModuleScalerBlockData as __packed -- which is fine.